src/test/java/org/medicmobile/webapp/mobile/FilePickerHandlerTest.java

```diff
@@ -0,0 +1,465 @@
+
+import androidx.core.content.FileProvider;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.MockedStatic;
+import org.mockito.MockitoAnnotations;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(sdk=28)
+public class FilePickerHandlerTest {
+	@Mock
+	Activity contextMock;
+
+	@Before
+	public void setup() {
+		MockitoAnnotations.openMocks(this);
+	}
+
+	private void stubMethodsInContextMock() {
+		String packageName = "org.medicmobile.webapp.mobile";
+		ApplicationInfo applicationInfo = mock(ApplicationInfo.class);
+		applicationInfo.packageName = packageName;
+
+		ActivityInfo activityInfo = mock(ActivityInfo.class);
+		activityInfo.applicationInfo = applicationInfo;
+		activityInfo.name = "TEST_ACTIVITY";
+
+		ResolveInfo resolveInfo = mock(ResolveInfo.class);
+		resolveInfo.activityInfo = activityInfo;
+
+		PackageManager packageManager = mock(PackageManager.class);
+		when(packageManager.resolveActivity(any(Intent.class), anyInt())).thenReturn(resolveInfo);
+
+		when(contextMock.getPackageManager()).thenReturn(packageManager);
+		when(contextMock.getExternalFilesDir(any())).thenReturn(mock(File.class));
+		when(contextMock.getPackageName()).thenReturn(packageName);
+	}
+
+	@Test
+	public void processResult_withResultOk_sendsFileUri() {
+		try (MockedStatic<MedicLog> medicLogMock = mockStatic(MedicLog.class)) {
+			//> GIVEN
+			ValueCallback<Uri[]> filePickerCallbackMock = mock(ValueCallback.class);
+			FilePickerHandler filePickerHandler = new FilePickerHandler(contextMock);
+			filePickerHandler.setFilePickerCallback(filePickerCallbackMock);
+
+			Uri uri = Uri.parse("content://some/content/image.jpg");
+			Intent intent = mock(Intent.class);
+			when(intent.getData()).thenReturn(uri);
+
+			//> WHEN
+			filePickerHandler.processResult(RESULT_OK, intent);
+
+			//> THEN
+			verify(filePickerCallbackMock).onReceiveValue(eq(new Uri[]{ uri }));
+			medicLogMock.verify(() -> MedicLog.trace(
+				eq(filePickerHandler),
+				eq("FilePickerHandler :: Sending data back to webapp, URI: %s"),
+				eq(uri)
+			));
+			medicLogMock.verify(() -> MedicLog.warn(
+				eq(filePickerHandler),
+				eq("FilePickerHandler :: Cannot send data back to webapp, filePickerCallback is null.")
+			), never());
+		}
+	}
+
+	@Test
+	public void processResult_withNullUri_finishesPicker() {
+		try (MockedStatic<MedicLog> medicLogMock = mockStatic(MedicLog.class)) {
+			//> GIVEN
+			ValueCallback<Uri[]> filePickerCallbackMock = mock(ValueCallback.class);
+			FilePickerHandler filePickerHandler = new FilePickerHandler(contextMock);
+			filePickerHandler.setFilePickerCallback(filePickerCallbackMock);
+
+			Intent intent = mock(Intent.class);
+			when(intent.getData()).thenReturn(null);
```

Comment:
I'm not sure what should happen in webapp at this point. If I cancel the upload, I get an invalid image now.

Reply:
🤔 I wonder if webapp can detect an invalid file and ignore it. I'll have a look

Reply:
Like when bytes are zero

Reply:
I've mitigated this one in the android app, but still needs the fix in webapp.